src/mame/sega/segaai.cpp

```diff
@@ -0,0 +1,822 @@
+    1 2           1,2,3 data lines
+  3 4   5         4 ??          5,8 data lines
+   6 7 8          6 GND         7 +5V
+
+
+*/
+
+#include "emu.h"
+
+#include "bus/segaai/segaai_exp.h"
+#include "bus/segaai/segaai_slot.h"
+#include "cpu/nec/nec.h"
+#include "imagedev/cassette.h"
+#include "machine/i8251.h"
+#include "machine/i8255.h"
+#include "sound/sn76496.h"
+#include "sound/upd7759.h"
+#include "video/v9938.h"
+
+#include "softlist.h"
+#include "softlist_dev.h"
+#include "speaker.h"
+
+#include "segaai.lh"
+
+#define VERBOSE (LOG_GENERAL)
+#include "logmacro.h"
+
+
+#define TOUCHPAD_ROWS 20
+
+
+class segaai_state : public driver_device
+{
+public:
+	segaai_state(const machine_config &mconfig, device_type type, const char *tag)
+		: driver_device(mconfig, type, tag)
+		, m_maincpu(*this, "maincpu")
+		, m_sound(*this, "sn76489a")
+		, m_v9938(*this, "v9938")
+		, m_upd7759(*this, "upd7759")
+		, m_cassette(*this, "cassette")
+		, m_cardslot(*this, "cardslot")
+		, m_expslot(*this, "exp")
+		, m_port4(*this, "PORT4")
+		, m_port5(*this, "PORT5")
+		, m_port_tp(*this, "TP.%u", 0)
+		, m_vector(0)
+	{ }
+
+	void segaai(machine_config &config);
+
+protected:
+	virtual void machine_start() override;
+
+private:
+	void vdp_interrupt(int state);
+	void upd7759_drq_w(int state);
+	IRQ_CALLBACK_MEMBER(irq_callback);
+	u8 i8255_porta_r();
+	u8 i8255_portb_r();
+	u8 i8255_portc_r();
+	void i8255_portc_w(u8 data);
+	void upd7759_ctrl_w(u8 data);
+	void upd7759_data_w(u8 data);
+	void port1c_w(u8 data);
+	void port1d_w(u8 data);
+	void port1e_w(u8 data);
+	u8 port1e_r();
+	u8 irq_enable_r();
+	void irq_enable_w(u8 data);
+	void irq_select_w(u8 data);
+
+	static constexpr u8 VECTOR_V9938 = 0xf8;
+	static constexpr u8 VECTOR_I8251_SEND = 0xf9;
+	static constexpr u8 VECTOR_I8251_RECEIVE = 0xfa;
+	static constexpr u8 VECTOR_UPD7759 = 0xfb;
+
+	static constexpr u8 IRQ_V9938 = 0x01;
+	static constexpr u8 IRQ_UPD7759 = 0x08;
+
+	// 8255 Port A bits
+	static constexpr u8 UPD7759_BUSY = 0x40;
+	// 8255 Port B bits
+	static constexpr u8 TOUCH_PAD_PRESSED = 0x02;
+	static constexpr u8 TOUCH_PAD_DATA_AVAILABLE = 0x04;
+
+	static constexpr u8 UPD7759_MODE = 0x01;
+	static constexpr u8 UPD7759_BANK = 0x02;
+
+	void mem_map(address_map &map);
+	void io_map(address_map &map);
+	void update_irq_state();
+	bool get_touchpad_pressed();
+	u32 get_vector() { return m_vector; }
+
+	required_device<cpu_device> m_maincpu;
+	required_device<sn76489a_device> m_sound;
+	required_device<v9938_device> m_v9938;
+	required_device<upd7759_device> m_upd7759;
+	required_device<cassette_image_device> m_cassette;
+	required_device<segaai_card_slot_device> m_cardslot;
+	required_device<segaai_exp_slot_device> m_expslot;
+	required_ioport m_port4;
+	required_ioport m_port5;
+	required_ioport_array<TOUCHPAD_ROWS> m_port_tp;
+
+	u8 m_i8255_portb;
+	u8 m_upd7759_ctrl;
+	u8 m_port_1c;
+	u8 m_port_1d;
+	u8 m_port_1e;
+	int m_prev_v9938_irq;
+	int m_prev_upd7759_irq;
```

Comment:
Please use explicitly sized types for things that are registered for save states.

Reply:
Changed to an explicitly sized type.